File: sdks/go/builderengine/billing.go
package builderengine

import (
	"context"
	"fmt"
)

// BillingService handles billing-related API calls
type BillingService struct {
	client *Client
}

// PaymentMethod represents a payment method
type PaymentMethod struct {
	ID        string `json:"id"`
	Type      string `json:"type"`
	Last4     string `json:"last4,omitempty"`
	Brand     string `json:"brand,omitempty"`
	ExpiryMonth int  `json:"expiry_month,omitempty"`
	ExpiryYear  int  `json:"expiry_year,omitempty"`
	IsDefault bool   `json:"is_default"`
}

// BillingPortalSession represents a billing portal session
type BillingPortalSession struct {
	URL string `json:"url"`
}

// GetSubscription returns the current subscription
func (s *BillingService) GetSubscription(ctx context.Context) (*Subscription, error) {
	var result Subscription
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   "/api/v1/billing/subscription",
	}, &result)
	return &result, err
}

// ListPlans returns available plans
func (s *BillingService) ListPlans(ctx context.Context) ([]map[string]interface{}, error) {
	var result struct {
		Data []map[string]interface{} `json:"data"`
	}
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   "/api/v1/billing/plans",
	}, &result)
	return result.Data, err
}

// ChangePlan changes the subscription plan
func (s *BillingService) ChangePlan(ctx context.Context, planID string) (*Subscription, error) {
	var result Subscription
	err := s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   "/api/v1/billing/subscription/change",
		Body:   map[string]string{"plan_id": planID},
	}, &result)
	return &result, err
}

// CancelSubscription cancels the subscription
func (s *BillingService) CancelSubscription(ctx context.Context, cancelAtPeriodEnd bool) (*Subscription, error) {
	var result Subscription
	err := s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   "/api/v1/billing/subscription/cancel",
		Body:   map[string]bool{"cancel_at_period_end": cancelAtPeriodEnd},
	}, &result)
	return &result, err
}

// ResumeSubscription resumes a canceled subscription
func (s *BillingService) ResumeSubscription(ctx context.Context) (*Subscription, error) {
	var result Subscription
	err := s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   "/api/v1/billing/subscription/resume",
	}, &result)
	return &result, err
}

// ListInvoices returns a list of invoices
func (s *BillingService) ListInvoices(ctx context.Context, params ListParams) (*PaginatedResponse[Invoice], error) {
	var result PaginatedResponse[Invoice]
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   "/api/v1/billing/invoices",
		Params: params.ToMap(),
	}, &result)
	return &result, err
}

// GetInvoice returns an invoice by ID
func (s *BillingService) GetInvoice(ctx context.Context, invoiceID string) (*Invoice, error) {
	var result Invoice
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   fmt.Sprintf("/api/v1/billing/invoices/%s", invoiceID),
	}, &result)
	return &result, err
}

// DownloadInvoice returns a download URL for an invoice
func (s *BillingService) DownloadInvoice(ctx context.Context, invoiceID string) (string, error) {
	var result struct {
		URL string `json:"url"`
	}
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   fmt.Sprintf("/api/v1/billing/invoices/%s/download", invoiceID),
	}, &result)
	return result.URL, err
}

// ListPaymentMethods returns payment methods
func (s *BillingService) ListPaymentMethods(ctx context.Context) ([]PaymentMethod, error) {
	var result struct {
		Data []PaymentMethod `json:"data"`
	}
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   "/api/v1/billing/payment-methods",
	}, &result)
	return result.Data, err
}

// AddPaymentMethod adds a new payment method
func (s *BillingService) AddPaymentMethod(ctx context.Context, setupIntentID string) (*PaymentMethod, error) {
	var result PaymentMethod
	err := s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   "/api/v1/billing/payment-methods",
		Body:   map[string]string{"setup_intent_id": setupIntentID},
	}, &result)
	return &result, err
}

// SetDefaultPaymentMethod sets the default payment method
func (s *BillingService) SetDefaultPaymentMethod(ctx context.Context, paymentMethodID string) error {
	return s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   fmt.Sprintf("/api/v1/billing/payment-methods/%s/default", paymentMethodID),
	}, nil)
}

// DeletePaymentMethod deletes a payment method
func (s *BillingService) DeletePaymentMethod(ctx context.Context, paymentMethodID string) error {
	return s.client.Request(ctx, RequestOptions{
		Method: "DELETE",
		Path:   fmt.Sprintf("/api/v1/billing/payment-methods/%s", paymentMethodID),
	}, nil)
}

// GetBalance returns the current credit balance
func (s *BillingService) GetBalance(ctx context.Context) (map[string]interface{}, error) {
	var result map[string]interface{}
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   "/api/v1/billing/balance",
	}, &result)
	return result, err
}

// AddCredits adds credits to the account
func (s *BillingService) AddCredits(ctx context.Context, amount float64) (map[string]interface{}, error) {
	var result map[string]interface{}
	err := s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   "/api/v1/billing/credits",
		Body:   map[string]float64{"amount": amount},
	}, &result)
	return result, err
}

// CreatePortalSession creates a billing portal session
func (s *BillingService) CreatePortalSession(ctx context.Context) (*BillingPortalSession, error) {
	var result BillingPortalSession
	err := s.client.Request(ctx, RequestOptions{
		Method: "POST",
		Path:   "/api/v1/billing/portal-session",
	}, &result)
	return &result, err
}

// GetUsageSummary returns usage summary for billing
func (s *BillingService) GetUsageSummary(ctx context.Context, period string) (map[string]interface{}, error) {
	var result map[string]interface{}
	params := map[string]string{}
	if period != "" {
		params["period"] = period
	}
	err := s.client.Request(ctx, RequestOptions{
		Method: "GET",
		Path:   "/api/v1/billing/usage",
		Params: params,
	}, &result)
	return result, err
}

File: sdks/python/builderengine/resources/billing.py
"""
Builder Engine Python SDK - Billing Resource

This module provides methods for managing billing and subscriptions.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional, Dict, Any, List

from builderengine.resources.base import BaseResource, PaginatedResponse
from builderengine.models import Subscription, Invoice, PaymentMethod, Usage
from builderengine.config import Endpoints

if TYPE_CHECKING:
    from builderengine.client import BuilderEngine


class BillingResource(BaseResource):
    """
    Resource for managing billing and subscriptions.

    This resource provides access to subscription management, invoices,
    payment methods, and usage tracking.

    Example:
        >>> client = BuilderEngine(api_key="...")
        >>> usage = client.billing.get_usage()
        >>> print(f"Minutes used: {usage.minutes_used}/{usage.minutes_limit}")
    """

    def get_subscription(self) -> Subscription:
        """
        Get the current subscription.

        Returns:
            Subscription object
        """
        response = self._get(Endpoints.BILLING_SUBSCRIPTION)
        return Subscription.from_dict(response)

    def get_plans(self) -> List[Dict[str, Any]]:
        """
        Get available subscription plans.

        Returns:
            List of plans with pricing and features
        """
        response = self._get(f"{Endpoints.BILLING_SUBSCRIPTION}/plans")
        return response.get("plans", [])

    def change_plan(
        self,
        plan_id: str,
        prorate: bool = True,
    ) -> Subscription:
        """
        Change subscription plan.

        Args:
            plan_id: ID of the new plan
            prorate: Whether to prorate the change

        Returns:
            Updated Subscription object

        Example:
            >>> subscription = client.billing.change_plan("plan_pro")
            >>> print(f"New plan: {subscription.plan_name}")
        """
        response = self._post(f"{Endpoints.BILLING_SUBSCRIPTION}/change", json={
            "plan_id": plan_id,
            "prorate": prorate,
        })
        return Subscription.from_dict(response)

    def cancel_subscription(
        self,
        at_period_end: bool = True,
        reason: Optional[str] = None,
    ) -> Subscription:
        """
        Cancel the subscription.

        Args:
            at_period_end: Cancel at end of billing period (vs immediately)
            reason: Reason for cancellation

        Returns:
            Updated Subscription object
        """
        data = {"at_period_end": at_period_end}
        if reason:
            data["reason"] = reason

        response = self._post(f"{Endpoints.BILLING_SUBSCRIPTION}/cancel", json=data)
        return Subscription.from_dict(response)

    def resume_subscription(self) -> Subscription:
        """
        Resume a canceled subscription.

        Returns:
            Updated Subscription object
        """
        response = self._post(f"{Endpoints.BILLING_SUBSCRIPTION}/resume")
        return Subscription.from_dict(response)

    # Usage

    def get_usage(
        self,
        period: str = "current",
    ) -> Usage:
        """
        Get usage for a billing period.

        Args:
            period: Billing period (current, previous, or date)

        Returns:
            Usage object with limits and current usage
        """
        response = self._get(Endpoints.BILLING_USAGE, params={"period": period})
        return Usage.from_dict(response)

    def get_usage_history(
        self,
        page: int = 1,
        page_size: int = 12,
    ) -> Dict[str, Any]:
        """
        Get usage history.

        Args:
            page: Page number
            page_size: Items per page

        Returns:
            Paginated usage history
        """
        return self._get(f"{Endpoints.BILLING_USAGE}/history", params={
            "page": page,
            "page_size": page_size,
        })

    # Invoices

    def list_invoices(
        self,
        page: int = 1,
        page_size: int = 20,
        status: Optional[str] = None,
    ) -> PaginatedResponse[Invoice]:
        """
        List invoices.

        Args:
            page: Page number
            page_size: Items per page
            status: Filter by status (draft, open, paid, void)

        Returns:
            PaginatedResponse containing Invoice objects
        """
        params = {
            "page": page,
            "page_size": page_size,
        }
        if status:
            params["status"] = status

        response = self._get(Endpoints.BILLING_INVOICES, params=params)
        return self._parse_paginated_response(response, Invoice)

    def get_invoice(self, invoice_id: str) -> Invoice:
        """
        Get an invoice by ID.

        Args:
            invoice_id: The invoice's unique identifier

        Returns:
            Invoice object
        """
        path = Endpoints.BILLING_INVOICE.format(invoice_id=invoice_id)
        response = self._get(path)
        return Invoice.from_dict(response)

    def download_invoice(self, invoice_id: str) -> Dict[str, Any]:
        """
        Get invoice PDF download URL.

        Args:
            invoice_id: The invoice's unique identifier

        Returns:
            Download URL
        """
        path = f"{Endpoints.BILLING_INVOICE.format(invoice_id=invoice_id)}/pdf"
        return self._get(path)

    def pay_invoice(self, invoice_id: str, payment_method_id: Optional[str] = None) -> Invoice:
        """
        Pay an open invoice.

        Args:
            invoice_id: The invoice's unique identifier
            payment_method_id: Payment method to use (default if not specified)

        Returns:
            Updated Invoice object
        """
        path = f"{Endpoints.BILLING_INVOICE.format(invoice_id=invoice_id)}/pay"
        data = {}
        if payment_method_id:
            data["payment_method_id"] = payment_method_id
        response = self._post(path, json=data)
        return Invoice.from_dict(response)

    # Payment Methods

    def list_payment_methods(self) -> List[PaymentMethod]:
        """
        List payment methods.

        Returns:
            List of PaymentMethod objects
        """
        response = self._get(Endpoints.BILLING_PAYMENT_METHODS)
        return [PaymentMethod.from_dict(pm) for pm in response.get("payment_methods", [])]

    def get_payment_method(self, payment_method_id: str) -> PaymentMethod:
        """
        Get a payment method by ID.

        Args:
            payment_method_id: The payment method's unique identifier

        Returns:
            PaymentMethod object
        """
        path = Endpoints.BILLING_PAYMENT_METHOD.format(payment_method_id=payment_method_id)
        response = self._get(path)
        return PaymentMethod.from_dict(response)

    def add_payment_method(
        self,
        type: str,
        token: str,
        set_default: bool = False,
        billing_address: Optional[Dict[str, str]] = None,
    ) -> PaymentMethod:
        """
        Add a payment method.

        Args:
            type: Payment method type (card, bank_account)
            token: Payment token from Stripe.js
            set_default: Set as default payment method
            billing_address: Billing address

        Returns:
            Created PaymentMethod object

        Example:
            >>> # Token from Stripe.js
            >>> pm = client.billing.add_payment_method(
            ...     type="card",
            ...     token="tok_visa",
            ...     set_default=True
            ... )
        """
        data = {
            "type": type,
            "token": token,
            "set_default": set_default,
        }
        if billing_address:
            data["billing_address"] = billing_address

        response = self._post(Endpoints.BILLING_PAYMENT_METHODS, json=data)
        return PaymentMethod.from_dict(response)

    def set_default_payment_method(self, payment_method_id: str) -> PaymentMethod:
        """
        Set a payment method as default.

        Args:
            payment_method_id: The payment method's unique identifier

        Returns:
            Updated PaymentMethod object
        """
        path = f"{Endpoints.BILLING_PAYMENT_METHOD.format(payment_method_id=payment_method_id)}/default"
        response = self._post(path)
        return PaymentMethod.from_dict(response)

    def delete_payment_method(self, payment_method_id: str) -> None:
        """
        Delete a payment method.

        Args:
            payment_method_id: The payment method's unique identifier
        """
        path = Endpoints.BILLING_PAYMENT_METHOD.format(payment_method_id=payment_method_id)
        self._delete(path)

    # Checkout and Portal

    def create_checkout_session(
        self,
        plan_id: str,
        success_url: str,
        cancel_url: str,
    ) -> Dict[str, Any]:
        """
        Create a Stripe checkout session.

        Args:
            plan_id: Plan to subscribe to
            success_url: URL to redirect on success
            cancel_url: URL to redirect on cancel

        Returns:
            Checkout session with URL
        """
        return self._post(Endpoints.BILLING_CHECKOUT, json={
            "plan_id": plan_id,
            "success_url": success_url,
            "cancel_url": cancel_url,
        })

    def create_portal_session(self, return_url: str) -> Dict[str, Any]:
        """
        Create a Stripe customer portal session.

        Args:
            return_url: URL to return to after portal

        Returns:
            Portal session with URL
        """
        return self._post(Endpoints.BILLING_PORTAL, json={"return_url": return_url})

    # Credits and Add-ons

    def get_credits(self) -> Dict[str, Any]:
        """
        Get available credits.

        Returns:
            Credit balance and history
        """
        return self._get(f"{Endpoints.BILLING_USAGE}/credits")

    def purchase_credits(
        self,
        amount: float,
        payment_method_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Purchase additional credits.

        Args:
            amount: Amount in dollars
            payment_method_id: Payment method to use

        Returns:
            Purchase confirmation
        """
        data = {"amount": amount}
        if payment_method_id:
            data["payment_method_id"] = payment_method_id

        return self._post(f"{Endpoints.BILLING_USAGE}/credits/purchase", json=data)

    def apply_promo_code(self, code: str) -> Dict[str, Any]:
        """
        Apply a promotional code.

        Args:
            code: Promo code to apply

        Returns:
            Result with discount details
        """
        return self._post(f"{Endpoints.BILLING_SUBSCRIPTION}/promo", json={"code": code})

    def get_spending_limit(self) -> Dict[str, Any]:
        """
        Get spending limit settings.

        Returns:
            Current spending limit configuration
        """
        return self._get(f"{Endpoints.BILLING_USAGE}/spending-limit")

    def set_spending_limit(
        self,
        limit: float,
        action: str = "notify",
    ) -> Dict[str, Any]:
        """
        Set spending limit.

        Args:
            limit: Monthly spending limit in dollars
            action: Action when limit reached (notify, pause, none)

        Returns:
            Updated spending limit configuration
        """
        return self._post(f"{Endpoints.BILLING_USAGE}/spending-limit", json={
            "limit": limit,
            "action": action,
        })
